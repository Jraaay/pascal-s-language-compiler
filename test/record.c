struct {
	char title[101]; 
	char author[50]; 
	char subject[100]; 
	int bookid; 
	struct {
		char title[101]; 
		char author[51]; 
		char subject[100]; 
		int bookid; 
	}
	Books2; 
}
Books; 
char a[100]; 
int b; 

int main(int argc, char* argv[]){
	a[0] = Books.title[0]; 
	a[0] = Books.Books2.title[0]; 
	a[0] = Books.Books2.author[b]; 
	a[0] = Books.Books2.author[10]; 
	b = Books.bookid; 
}
