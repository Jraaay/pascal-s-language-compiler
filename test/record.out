{
    "ast": {
        "length": 5,
        "type": "programstruct",
        "program_head": {
            "length": 6,
            "type": "program_head",
            "ID": "example",
            "idlist": {
                "length": 4,
                "type": "idlist",
                "ids": [
                    "input",
                    "output"
                ]
            }
        },
        "program_body": {
            "length": 5,
            "type": "program_body",
            "const_declarations": null,
            "var_declarations": {
                "length": 4,
                "type": "var_declarations",
                "var_declaration": {
                    "length": 6,
                    "type": "var_declaration",
                    "values": [
                        {
                            "idlist": {
                                "length": 2,
                                "type": "idlist",
                                "ids": [
                                    "Books"
                                ]
                            },
                            "type": {
                                "length": 4,
                                "type": "type",
                                "_type": "RECORD",
                                "multype": {
                                    "length": 6,
                                    "id": 3,
                                    "type": [
                                        {
                                            "length": 7,
                                            "type": "type",
                                            "_type": "ARRAY",
                                            "period": {
                                                "length": 4,
                                                "type": "period",
                                                "values": [
                                                    {
                                                        "start": 1,
                                                        "END": 50
                                                    }
                                                ]
                                            },
                                            "basic_type": {
                                                "type": "basic_type",
                                                "_type": "CHAR"
                                            }
                                        },
                                        {
                                            "length": 7,
                                            "type": "type",
                                            "_type": "ARRAY",
                                            "period": {
                                                "length": 4,
                                                "type": "period",
                                                "values": [
                                                    {
                                                        "start": 1,
                                                        "END": 50
                                                    }
                                                ]
                                            },
                                            "basic_type": {
                                                "type": "basic_type",
                                                "_type": "CHAR"
                                            }
                                        },
                                        {
                                            "length": 7,
                                            "type": "type",
                                            "_type": "ARRAY",
                                            "period": {
                                                "length": 4,
                                                "type": "period",
                                                "values": [
                                                    {
                                                        "start": 1,
                                                        "END": 100
                                                    }
                                                ]
                                            },
                                            "basic_type": {
                                                "type": "basic_type",
                                                "_type": "CHAR"
                                            }
                                        },
                                        {
                                            "length": 2,
                                            "type": "type",
                                            "_type": {
                                                "type": "basic_type",
                                                "_type": "INTEGER"
                                            }
                                        }
                                    ],
                                    "multype": {
                                        "length": 6,
                                        "id": 2,
                                        "type": [
                                            {
                                                "length": 7,
                                                "type": "type",
                                                "_type": "ARRAY",
                                                "period": {
                                                    "length": 4,
                                                    "type": "period",
                                                    "values": [
                                                        {
                                                            "start": 1,
                                                            "END": 50
                                                        }
                                                    ]
                                                },
                                                "basic_type": {
                                                    "type": "basic_type",
                                                    "_type": "CHAR"
                                                }
                                            },
                                            {
                                                "length": 7,
                                                "type": "type",
                                                "_type": "ARRAY",
                                                "period": {
                                                    "length": 4,
                                                    "type": "period",
                                                    "values": [
                                                        {
                                                            "start": 1,
                                                            "END": 50
                                                        }
                                                    ]
                                                },
                                                "basic_type": {
                                                    "type": "basic_type",
                                                    "_type": "CHAR"
                                                }
                                            },
                                            {
                                                "length": 7,
                                                "type": "type",
                                                "_type": "ARRAY",
                                                "period": {
                                                    "length": 4,
                                                    "type": "period",
                                                    "values": [
                                                        {
                                                            "start": 1,
                                                            "END": 100
                                                        }
                                                    ]
                                                },
                                                "basic_type": {
                                                    "type": "basic_type",
                                                    "_type": "CHAR"
                                                }
                                            }
                                        ],
                                        "multype": {
                                            "length": 6,
                                            "id": 1,
                                            "type": [
                                                {
                                                    "length": 7,
                                                    "type": "type",
                                                    "_type": "ARRAY",
                                                    "period": {
                                                        "length": 4,
                                                        "type": "period",
                                                        "values": [
                                                            {
                                                                "start": 1,
                                                                "END": 50
                                                            }
                                                        ]
                                                    },
                                                    "basic_type": {
                                                        "type": "basic_type",
                                                        "_type": "CHAR"
                                                    }
                                                },
                                                {
                                                    "length": 7,
                                                    "type": "type",
                                                    "_type": "ARRAY",
                                                    "period": {
                                                        "length": 4,
                                                        "type": "period",
                                                        "values": [
                                                            {
                                                                "start": 1,
                                                                "END": 50
                                                            }
                                                        ]
                                                    },
                                                    "basic_type": {
                                                        "type": "basic_type",
                                                        "_type": "CHAR"
                                                    }
                                                }
                                            ],
                                            "multype": {
                                                "length": 5,
                                                "id": 0,
                                                "type": [
                                                    {
                                                        "length": 7,
                                                        "type": "type",
                                                        "_type": "ARRAY",
                                                        "period": {
                                                            "length": 4,
                                                            "type": "period",
                                                            "values": [
                                                                {
                                                                    "start": 1,
                                                                    "END": 50
                                                                }
                                                            ]
                                                        },
                                                        "basic_type": {
                                                            "type": "basic_type",
                                                            "_type": "CHAR"
                                                        }
                                                    }
                                                ],
                                                "ID": [
                                                    "title"
                                                ]
                                            },
                                            "ID": [
                                                "title",
                                                "author"
                                            ]
                                        },
                                        "ID": [
                                            "title",
                                            "author",
                                            "subject"
                                        ]
                                    },
                                    "ID": [
                                        "title",
                                        "author",
                                        "subject",
                                        "bookid"
                                    ]
                                }
                            }
                        },
                        {
                            "idlist": {
                                "length": 2,
                                "type": "idlist",
                                "ids": [
                                    "a"
                                ]
                            },
                            "type": {
                                "length": 7,
                                "type": "type",
                                "_type": "ARRAY",
                                "period": {
                                    "length": 4,
                                    "type": "period",
                                    "values": [
                                        {
                                            "start": 1,
                                            "END": 100
                                        }
                                    ]
                                },
                                "basic_type": {
                                    "type": "basic_type",
                                    "_type": "CHAR"
                                }
                            }
                        },
                        {
                            "idlist": {
                                "length": 2,
                                "type": "idlist",
                                "ids": [
                                    "b"
                                ]
                            },
                            "type": {
                                "length": 2,
                                "type": "type",
                                "_type": {
                                    "type": "basic_type",
                                    "_type": "INTEGER"
                                }
                            }
                        }
                    ]
                }
            },
            "subprogram_declarations": {
                "length": 1,
                "type": "subprogram_declarations",
                "subprograms": []
            },
            "compound_statement": {
                "length": 4,
                "type": "compound_statement",
                "statement_list": {
                    "length": 4,
                    "type": "statement_list",
                    "statements": [
                        {
                            "length": 4,
                            "type": "statement",
                            "_type": "variable",
                            "variable": {
                                "length": 3,
                                "type": "variable",
                                "__type": "CHAR",
                                "ID": "a",
                                "id_varpart": {
                                    "length": 4,
                                    "type": "id_varpart",
                                    "expression_list": {
                                        "length": 2,
                                        "type": "expression_list",
                                        "__type": [
                                            "INTEGER"
                                        ],
                                        "expressions": [
                                            {
                                                "length": 2,
                                                "type": "expression",
                                                "__type": "INTEGER",
                                                "simple_expression": {
                                                    "length": 2,
                                                    "type": "simple_expression",
                                                    "__type": "INTEGER",
                                                    "term": {
                                                        "length": 2,
                                                        "type": "term",
                                                        "__type": "INTEGER",
                                                        "factor": {
                                                            "length": 2,
                                                            "type": "factor",
                                                            "_type": "NUM",
                                                            "__type": "INTEGER",
                                                            "NUM": 1
                                                        }
                                                    }
                                                }
                                            }
                                        ]
                                    }
                                }
                            },
                            "ASSIGNOP": ":=",
                            "expression": {
                                "length": 2,
                                "type": "expression",
                                "__type": "CHAR",
                                "simple_expression": {
                                    "length": 2,
                                    "type": "simple_expression",
                                    "__type": "CHAR",
                                    "term": {
                                        "length": 2,
                                        "type": "term",
                                        "__type": "CHAR",
                                        "factor": {
                                            "length": 2,
                                            "type": "factor",
                                            "_type": "variable",
                                            "__type": "CHAR",
                                            "variable": {
                                                "length": 3,
                                                "type": "variable",
                                                "__type": "CHAR",
                                                "ID": [
                                                    "Books",
                                                    "title"
                                                ],
                                                "id_varpart": {
                                                    "length": 4,
                                                    "type": "id_varpart",
                                                    "expression_list": {
                                                        "length": 2,
                                                        "type": "expression_list",
                                                        "__type": [
                                                            "INTEGER"
                                                        ],
                                                        "expressions": [
                                                            {
                                                                "length": 2,
                                                                "type": "expression",
                                                                "__type": "INTEGER",
                                                                "simple_expression": {
                                                                    "length": 2,
                                                                    "type": "simple_expression",
                                                                    "__type": "INTEGER",
                                                                    "term": {
                                                                        "length": 2,
                                                                        "type": "term",
                                                                        "__type": "INTEGER",
                                                                        "factor": {
                                                                            "length": 2,
                                                                            "type": "factor",
                                                                            "_type": "NUM",
                                                                            "__type": "INTEGER",
                                                                            "NUM": 1
                                                                        }
                                                                    }
                                                                }
                                                            }
                                                        ]
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        },
                        {
                            "length": 4,
                            "type": "statement",
                            "_type": "variable",
                            "variable": {
                                "length": 3,
                                "type": "variable",
                                "__type": "INTEGER",
                                "ID": "b",
                                "id_varpart": {
                                    "length": 1,
                                    "type": "id_varpart",
                                    "expression_list": null
                                }
                            },
                            "ASSIGNOP": ":=",
                            "expression": {
                                "length": 2,
                                "type": "expression",
                                "__type": "INTEGER",
                                "simple_expression": {
                                    "length": 2,
                                    "type": "simple_expression",
                                    "__type": "INTEGER",
                                    "term": {
                                        "length": 2,
                                        "type": "term",
                                        "__type": "INTEGER",
                                        "factor": {
                                            "length": 2,
                                            "type": "factor",
                                            "_type": "variable",
                                            "__type": "INTEGER",
                                            "variable": {
                                                "length": 3,
                                                "type": "variable",
                                                "__type": "INTEGER",
                                                "ID": [
                                                    "Books",
                                                    "bookid"
                                                ],
                                                "id_varpart": {
                                                    "length": 1,
                                                    "type": "id_varpart",
                                                    "expression_list": null
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    ]
                }
            }
        }
    },
    "symbolTable": {
        "constants": [],
        "variables": [
            {
                "id": 4,
                "token": "Books",
                "type": "RECORD",
                "isArray": false,
                "dimension": 0,
                "size": [],
                "start": [],
                "recordTable": {
                    "variables": [
                        {
                            "id": 0,
                            "token": "title",
                            "type": "CHAR",
                            "isArray": true,
                            "dimension": 1,
                            "size": [
                                50
                            ],
                            "start": [
                                1
                            ],
                            "recordTable": null
                        },
                        {
                            "id": 1,
                            "token": "author",
                            "type": "CHAR",
                            "isArray": true,
                            "dimension": 1,
                            "size": [
                                50
                            ],
                            "start": [
                                1
                            ],
                            "recordTable": null
                        },
                        {
                            "id": 2,
                            "token": "subject",
                            "type": "CHAR",
                            "isArray": true,
                            "dimension": 1,
                            "size": [
                                100
                            ],
                            "start": [
                                1
                            ],
                            "recordTable": null
                        },
                        {
                            "id": 3,
                            "token": "bookid",
                            "type": "INTEGER",
                            "isArray": false,
                            "dimension": 0,
                            "size": [],
                            "start": [],
                            "recordTable": null
                        }
                    ]
                }
            },
            {
                "id": 5,
                "token": "a",
                "type": "CHAR",
                "isArray": true,
                "dimension": 1,
                "size": [
                    100
                ],
                "start": [
                    1
                ],
                "recordTable": null
            },
            {
                "id": 6,
                "token": "b",
                "type": "INTEGER",
                "isArray": false,
                "dimension": 0,
                "size": [],
                "start": [],
                "recordTable": null
            }
        ],
        "subFunc": []
    },
    "error": [],
    "warning": []
}